### ISSUE TITLE
[BUG] Failing rule tests with examples having CRLF line endings

### ISSUE THREAD

**Issue by roman-petrov** [dart-code-checker/dart-code-metrics#329]
* **Dart code metrics version:** 3.2.2
* **Dart sdk version:** 2.14.0-18.0.dev

The problem can be reproduced on almost any rule test: just convert line endings to CRLF in `example.dart`. 

I found this issue when I cloned the repository on my Windows machine at work. In my global `git` settings I have [core.autocrlf](https://docs.github.com/en/github/getting-started-with-github/configuring-git-to-handle-line-endings#global-settings-for-line-endings) enabled. So I got all text files in my working copy with CRLF line endings.

There is an offset verification mechanism in rule tests like
```
 RuleTestHelper.verifyIssues(
        issues: issues,
        startOffsets: [15, 116, 179, 288, 338, 387],
```

As I understand these offsets will be different for different line endings in `example.dart` so I got failed tests on my machine. I managed to get test passed after I manually converted line endings to LF in `example.dart` .

I see a number of solutions for this problem and I need your help to choose a best one:
1. Do nothing: contributors have to make sure that they clone repository with LF line endings. Maybe we should update contributing guide
2. Add `.gitattributes`(https://docs.github.com/en/github/getting-started-with-github/configuring-git-to-handle-line-endings#per-repository-settings) into the repository with `text eol=lf` option. **This approach has one drawback**: new `example.dart` files still might be created with CRLF line endings. So contributor might get passed tests locally, but failing tests on 
3. When loading examples into tests we can automatically convert line endings to LF. We can implement this change in [FileResolver.resolve](https://github.com/dart-code-checker/dart-code-metrics/blob/master/test/helpers/file_resolver.dart#L20) test helper method. IMHO this is most reliable approach.

What is your opinion on this?

**Comment by incendial** [comment]
Oh, that's an interesting problem.

> When loading examples into tests we can automatically convert line endings to LF. We can implement this change in FileResolver.resolve test helper method. IMHO this is most reliable approach.

This one looks very good, but I don't see a way to do that. That's how we create and resolve files right now:
```dart
final path = normalize(File(filePath).absolute.path);
final sourceUrl = Uri.parse(path);

// ignore: deprecated_member_use
final parseResult = await resolveFile(path: path) as ResolvedUnitResult;
```
As you can see, it uses only a file path to create a unit. How you want to fix CRLF's with only the path being used?

**Comment by incendial** [comment]
> files still might be created with LF line endings

Is it a typo and you meant CRLF?

**Comment by roman-petrov** [comment]
> > files still might be created with LF line endings
> 
> Is it a typo and you meant CRLF?

Sure, it's a typo; I updated issue text.

> As you can see, it uses only a file path to create a unit. How you want to fix CRLF's with only the path being used?

At the moment I don't have good understanding of project architecture. I thought that in [FileResolver.resolve](https://github.com/dart-code-checker/dart-code-metrics/blob/master/test/helpers/file_resolver.dart#L20) method we can just normalize line endings to LF for `parseResult.content` before returning it inside `InternalResolvedUnitResult`...

Generally, the idea is to always load all example files in tests with normalized line endings, but I am not completely sure about amount of work this will require to implement.

**Comment by incendial** [comment]
> we can just normalize line endings to LF for parseResult.content

It works in a bit different way - we pass a file to `resolveFile` function and get a `ResolvedUnitResult`. Then we create a wrapper for that in order to remove non-null assertions in our codebase. And all rules are using compilation unit. So, if we change `parseResult.content` it will not affect the unit itself.

So, I like the idea, but I don't see how we can do that without creating some kind of a temporary file with replaced CRLF.

**Comment by roman-petrov** [comment]
@incendial, thank you for explanations. I took a deeper look inside the project, now I see that my proposed solution to normalize line endings before Dart compiler parses source code is quite hard to implement.

So I created a PR https://github.com/dart-code-checker/dart-code-metrics/pull/331 with other proposed solution for this issue, please take a look.

**Comment by roman-petrov** [comment]
I tested #331 locally and everything seems to be working as expected. Please take a look let me know your opinion about it.

**Comment by incendial** [comment]
I'm kinda on a small vacation right now, so I'll review your other request first and then return to this one. Sorry for the delay.

**Comment by roman-petrov** [comment]
@incendial that's ok, take a good holiday!

**Comment by incendial** [comment]
> @incendial that's ok, take a good holiday!

Thank you! I'm finally back, let's continue discussion in PR.
